core_lib/src/tool/movetool.cpp

```diff
@@ -161,6 +162,8 @@ void MoveTool::pointerMoveEvent(PointerEvent* event)
 
 void MoveTool::pointerReleaseEvent(PointerEvent*)
 {
+    mEditor->backups()->backup(BackupType::SELECTION);
```

Comment:
I recently added a check to make sure it happens on only the new backup system. Was this comment written before that?

Or do you mean that we should check  for nullptrs because it's possible to call backup prior to calling savestates and crash?

Reply:
I'm thinking about solving it like this:

We move the member variables into a struct for convenience, called UndoSaveState, the state will only be kept while making a backup, so as soon as we add it to the stack, we'll clear it again, then we might only need to check whether the mSaveState exists.

Something like:

Implementation notes:
```
struct UndoSaveState {
	int mLayerId = 0;
    Layer::LAYER_TYPE mType = Layer::UNDEFINED;
    KeyFrame* mKeyframe = nullptr;

    QRectF mSelectionRect;
    qreal mSelectionRotationAngle = 0.0;
    qreal mSelectionScaleX = 0.0;
    qreal mSelectionScaleY = 0.0;
    QPointF mSelectionTranslation;
    QPointF mSelectionAnchor;
};

UndoSaveState BackupManager::saveStates()

....

BackupManager->backup(BackupType::STROKE)

BackupManager::backup(BackupType type)
{
    if (mSaveState == nullptr) { return }

    switch (type) { ... }
}

BackupManager::bitmap(const UndoSaveState* state, description)
{
    // Clear save state again after adding it to the stack
    mUndoState = nullptr
}

Another way could be to only create the struct and let the caller handle keep that data until they feed it as an argument, but I think it's safer to keep as much state as possible in the manager itself.

```

Reply:
My reason for following through with the alternative was not so much about exposing state, it was more about not being allowed to keep old state around.

In the proposed implementation, calling add(...) will throw an assert in debug builds when calling add(...) without having called saveStates(), because that's technically a bug, but it will do nothing in the actual release. The old state is invalidated once the command has been added to the stack, so which ensure that you can't use it more than once.

In the ideal scenario you would be able to call saveState prior to calling add(...) and that would indeed make the control flow easier to understand. In some situations however, for example with the select or move tool, you need to save the state on one event, store it somewhere and add it later in another event, eg. PointerPress -> saveStates() -> PointerRelease -> add(..)

Now before we input the saveState into add(...), we have to ensure that it's also valid state and that's up to the respective tool or operation using undo/redo. 

I've been experimenting with the concept and it works it some scenarios and complicates it others. 

> could invalidate the state on use by setting layerId = -1 or similar. Though I’m curious what other safety concerns you may have in mind.

Hmm true, we could invalidate the state right after but then we need to check for invalid state before calling add. I feel like this pattern will lead to less maintainable code in the end or what do you think?

Reply:
Sorry for the delay, I got side tracked a bit because of all the activity lately 🚀 

I came up with a pattern I think works well, I ended up using shared_ptr instead of unique_ptr, since there's technically nothing wrong with saving state multiple times, in different contexts. 

Let me know what you think 😃